File: core/workers/discogs.py
"""
discogs.py — Discogs enrichment worker.

Rate limit: 25 calls/min zonder token, 60/min met token.
We gebruiken 1 call per 2.5 sec als veilige marge.
"""

import logging
import time
from typing import Any, Optional

import discogs_client

from core.workers.base import BaseWorker

RATE_INTERVAL = 2.5   # seconden


class DiscogsWorker(BaseWorker):
    source = "discogs"

    def __init__(self, database, logger: logging.Logger,
                 token: Optional[str] = None,
                 user_agent: str = "LastfmMuziekApp/1.0 +https://github.com/muziek"):
        super().__init__(database, logger)
        self._last_call = 0.0
        self._token     = token
        self._ua        = user_agent

        self._client = discogs_client.Client(
            self._ua,
            user_token=token or "",
        )

    def _rate_limit(self) -> None:
        wait = max(0.0, RATE_INTERVAL - (time.time() - self._last_call))
        if wait > 0:
            time.sleep(wait)
        self._last_call = time.time()

    def process(self, item: dict) -> dict[str, Any]:
        try:
            data = self._search(item["entity_name"], item["entity_type"])
            if not data:
                return {"ok": False, "error": "No results found"}
            return {"ok": True, "data": data}
        except Exception as exc:
            self.log.warning(f"Discogs worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}

    def _search(self, name: str, entity_type: str = "artist") -> Optional[dict]:
        discogs_type = (
            "artist"  if entity_type == "artist" else
            "master"  if entity_type == "album"  else
            "release"
        )

        self._rate_limit()
        results = self._client.search(name, type=discogs_type)
        items   = list(results.page(1)[:5]) if results else []
        if not items:
            return None

        # Probeer exacte match op title
        norm  = lambda s: (s or "").lower().strip()
        exact = next((i for i in items if norm(getattr(i, "title", "")) == norm(name)), None)
        best  = exact or items[0]

        if entity_type == "artist":
            return self._enrich_artist(best, name)
        return self._map_release(best)

    def _enrich_artist(self, item, original_name: str) -> dict:
        base = {
            "discogsId":   item.id,
            "title":       item.title,
            "thumb":       item.thumb or None,
            "coverImage":  getattr(item, "cover_image", None),
            "uri":         item.url if hasattr(item, "url") else None,
            "resourceUrl": item.data.get("resource_url") if hasattr(item, "data") else None,
            "source":      "discogs",
            "fetchedAt":   int(time.time() * 1000),
        }

        # Haal detail op voor biografie etc.
        try:
            self._rate_limit()
            detail = item.fetch()
            base["profile"]        = getattr(detail, "profile", None)
            base["aliases"]        = [a.name for a in (getattr(detail, "aliases", None) or [])]
            base["members"]        = [m.name for m in (getattr(detail, "members", None) or [])]
            base["urls"]           = list(getattr(detail, "urls", []) or [])
            base["nameVariations"] = list(getattr(detail, "namevariations", []) or [])
        except Exception as exc:
            self.log.debug(f"Discogs artist detail mislukt voor '{original_name}': {exc}")

        return base

    def _map_release(self, item) -> dict:
        d = item.data if hasattr(item, "data") else {}
        return {
            "discogsId":  item.id,
            "title":      item.title,
            "year":       d.get("year"),
            "label":      ", ".join(d.get("label", [])) or None,
            "catno":      d.get("catno"),
            "format":     ", ".join(d.get("format", [])) or None,
            "genre":      d.get("genre", []),
            "style":      d.get("style", []),
            "country":    d.get("country"),
            "thumb":      item.thumb or None,
            "coverImage": d.get("cover_image"),
            "uri":        item.url if hasattr(item, "url") else None,
            "source":     "discogs",
            "fetchedAt":  int(time.time() * 1000),
        }

File: .claude/worktrees/vibrant-pike-750673/services/enrichment/workers/discogs.js
'use strict';
// ── Discogs worker ────────────────────────────────────────────────────────────
// API: https://api.discogs.com/database/search
// Rate limit: 25 calls/minuut zonder token, 60/min met user-agent token
// We gebruiken 1 call per 2.5 sec als veilige marge

const DISCOGS_BASE   = 'https://api.discogs.com';
const RATE_INTERVAL  = 2500; // ms

class DiscogsWorker {
  constructor(db, log, { token, userAgent } = {}) {
    this.db        = db;
    this.log       = log.child ? log.child({ worker: 'discogs' }) : log;
    this.token     = token     || null;
    this.userAgent = userAgent || 'LastfmMuziekApp/1.0 +https://github.com/muziek';
    this._lastCall = 0;
  }

  async _rateLimit() {
    const wait = Math.max(0, RATE_INTERVAL - (Date.now() - this._lastCall));
    if (wait > 0) await new Promise(r => setTimeout(r, wait));
    this._lastCall = Date.now();
  }

  _headers() {
    const h = { 'User-Agent': this.userAgent, 'Accept': 'application/json' };
    if (this.token) h['Authorization'] = `Discogs token=${this.token}`;
    return h;
  }

  async _get(url) {
    await this._rateLimit();
    const res = await fetch(url, {
      headers: this._headers(),
      signal:  AbortSignal.timeout(12_000),
    });
    if (res.status === 429) throw new Error('Discogs rate limit hit');
    if (!res.ok) throw new Error(`Discogs HTTP ${res.status} — ${url}`);
    return res.json();
  }

  async process(entity) {
    try {
      const data = await this.search(entity.entity_name, entity.entity_type);
      if (!data) return { ok: false, error: 'No results found' };
      return { ok: true, data };
    } catch (err) {
      this.log.warn({ err: err.message, entity: entity.entity_name }, 'Discogs fetch failed');
      return { ok: false, error: err.message };
    }
  }

  async search(name, entityType = 'artist') {
    const discogsType = entityType === 'artist' ? 'artist'
      : entityType === 'album' ? 'master'
      : 'release';

    const url = `${DISCOGS_BASE}/database/search?${new URLSearchParams({
      q:    name,
      type: discogsType,
      per_page: '5',
    })}`;

    const data = await this._get(url);
    const results = (data.results || []);
    if (!results.length) return null;

    const norm  = s => (s || '').toLowerCase().trim();
    const exact = results.find(r => norm(r.title) === norm(name));
    const best  = exact || results[0];

    if (entityType === 'artist') {
      return await this._enrichArtist(best, name);
    }
    return this._mapRelease(best);
  }

  async _enrichArtist(item, originalName) {
    const base = {
      discogsId:   item.id,
      title:       item.title,
      thumb:       item.thumb || null,
      coverImage:  item.cover_image || null,
      uri:         item.uri || null,
      resourceUrl: item.resource_url || null,
      source:      'discogs',
      fetchedAt:   Date.now(),
    };

    // Probeer artist-detail op te halen voor biografie en labels
    if (item.resource_url) {
      try {
        const detail = await this._get(item.resource_url);
        base.profile  = detail.profile  || null;  // biografie
        base.aliases  = (detail.aliases  || []).map(a => a.name);
        base.members  = (detail.members  || []).map(m => m.name);
        base.urls     = detail.urls || [];
        base.nameVariations = detail.namevariations || [];
      } catch (err) {
        this.log.debug({ err: err.message, name: originalName }, 'Discogs artist detail fetch failed');
      }
    }

    return base;
  }

  _mapRelease(item) {
    return {
      discogsId:   item.id,
      title:       item.title,
      year:        item.year || null,
      label:       (item.label || []).join(', ') || null,
      catno:       item.catno || null,
      format:      (item.format || []).join(', ') || null,
      genre:       item.genre || [],
      style:       item.style || [],
      country:     item.country || null,
      thumb:       item.thumb || null,
      coverImage:  item.cover_image || null,
      uri:         item.uri || null,
      source:      'discogs',
      fetchedAt:   Date.now(),
    };
  }
}

module.exports = { DiscogsWorker };
